### PR TITLE
feat: expose new fix api

### PR DESCRIPTION
**Pull request by slavik-lvovsky**

### CODE REVIEW

File: packages/npm-dependencies-validation/src/depIssuesFinder.ts [hunk 1]
```diff
@@ -42,6 +40,21 @@ export async function findDependencyIssues(
     devDepsCommandConfig
   );
   /* istanbul ignore next  -- inconsistent behavior under npm6-8 so not all branches are reachable here */
```
Comment (bd82) [inline packages/npm-dependencies-validation/src/depIssuesFinder.ts:42]: you can remove this coverage comment as well now

File: packages/npm-dependencies-validation/src/depIssuesFixer.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  print(internal.fixing(filePath), outputChannel);
+
+  const config = { commandArgs: ["install"], cwd };
+  await invokeNPMCommand(config, outputChannel);
+
+  print(internal.doneFixing(filePath), outputChannel);
+}
+
+function getDateAndTime(): string {
```
Comment (bd82) [inline packages/npm-dependencies-validation/src/depIssuesFixer.ts:30]: 1. descriptive name, for example look at: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString

2. Why do we need the full date here? (e.g including year/month/day)

File: packages/npm-dependencies-validation/src/logger.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { OutputChannel } from "./types";
+
+export function print(data: string, outputChannel?: OutputChannel): void {
+  if (outputChannel) {
+    outputChannel.append(data);
```
Comment (bd82) [inline packages/npm-dependencies-validation/src/logger.ts:5]: would it be safer to use `appendLine`

File: packages/npm-dependencies-validation/src/utils/fileUtil.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+export function toJsonObject<T>(jsonContent: string): T {
+  try {
+    const jsonObj: T = JSON.parse(jsonContent);
+    return jsonObj;
+  } catch (error) {
+    return emptyJsonObject<T>();
+  }
+}
+
+export async function isPathExist(absPath: string): Promise<boolean> {
```
Comment (bd82) [inline packages/npm-dependencies-validation/src/utils/fileUtil.ts:14]: From English grammar perspective I believe it should be: "doesPathExist"

File: packages/npm-dependencies-validation/test/utils/fileUtil.spec.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { expect } from "chai";
+import { toJsonObject } from "../../src/utils/fileUtil";
+
+describe("fileUtil unit tests", () => {
+  context("toJsonObject()", () => {
+    it("will return json object for valid json string", () => {
+      const content = `{"name": "test"}`;
+      type TestJson = { name: string };
+      const result: TestJson = toJsonObject<TestJson>(content);
+      expect(result).to.haveOwnProperty("name");
```
Comment (bd82) [inline packages/npm-dependencies-validation/test/utils/fileUtil.spec.ts:10]: not really important and also subjective.
But in such cases you can use `to.deep.equal({name: "test"})` and avoid the 
need for:
1. helper type definition
2. multiple assertions

File: packages/npm-dependencies-validation/test/utils/fileUtil.spec.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      expect(result.name).to.equal("test");
+    });
+
+    it("will return empty object for invalid json string", () => {
+      const content = `{"invalid json"}`;
+      const result: { name: string } = toJsonObject(content);
+      expect(result).to.be.empty;
```
Comment (bd82) [inline packages/npm-dependencies-validation/test/utils/fileUtil.spec.ts:17]: `to.be.empty` may not be the strongest assertion here, e.g: `[]` would also satisfy the constraint. 

`to.deep.equal({})` is probably the easiest way here.

File: packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    }
+  });
+
+  it("will not fix missing deps", async () => {
+    await expect(fixDependencyIssues("non_existing_package_json_path")).to.be
+      .fulfilled;
+  });
+
+  it("will fix missing deps", async function () {
+    this.timeout(20000);
```
Comment (bd82) [inline packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts:37]: use `npmSpawnTestTimeout` variable in config.ts

Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts:37]: 8000ms is not enough for this test

File: packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import { OutputChannel } from "../src/types";
+import { fixDependencyIssues, findDependencyIssues } from "../src/api";
+import { isPathExist } from "../src/utils/fileUtil";
+
+describe("`fixDependencyIssues()` function ", function () {
+  const outputChannel: OutputChannel = {
+    append: (data: string) => console.log(data),
+  };
+
+  const packageJsonPath = resolve(
```
Comment (bd82) [inline packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts:13]: these two `consts` are only used in **one** of the tests
so it is a little strange that they are at the describe level.

Also the `afterEach` is only needed for one of the tests.
unless you intend to add more tests in the future, perhaps the structure should be modified?

**Review by bd82 — COMMENTED** [review]
minor comments, see - in-line

File: packages/npm-dependencies-validation/src/depIssuesFinder.ts (hunk 1 above)
Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/src/depIssuesFinder.ts:42]: Done

File: packages/npm-dependencies-validation/src/depIssuesFixer.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import { isPathExist } from "./utils/fileUtil";
+import { invokeNPMCommand } from "./utils/npmUtil";
+import { print } from "./logger";
+import { createPackageJsonPaths } from "./utils/packageJsonUtil";
+
+export const internal = {
+  fixing: (absPath: string) =>
+    `\n${absPath}\n[${getDateAndTime()}] Fixing dependency issues...\n`,
+  doneFixing: (absPath: string) =>
+    `\n[${getDateAndTime()}] Done. \n${absPath}\n`,
```
[the remark this answers is not in the document]
Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/src/depIssuesFixer.ts:11]: Done

File: packages/npm-dependencies-validation/src/depIssuesFixer.ts (hunk 2 above)
Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/src/depIssuesFixer.ts:30]: Done

File: packages/npm-dependencies-validation/src/logger.ts (hunk 3 above)
Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/src/logger.ts:5]: Done

File: packages/npm-dependencies-validation/src/utils/fileUtil.ts (hunk 4 above)
Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/src/utils/fileUtil.ts:14]: Done

File: packages/npm-dependencies-validation/test/utils/fileUtil.spec.ts (hunk 5 above)
Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/test/utils/fileUtil.spec.ts:10]: Done

File: packages/npm-dependencies-validation/test/utils/fileUtil.spec.ts (hunk 6 above)
Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/test/utils/fileUtil.spec.ts:17]: Done

File: packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts (hunk 8 above)
Comment (slavik-lvovsky) [reply packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts]: Done

File: packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts (hunk 7 above)
Comment (bd82) [reply packages/npm-dependencies-validation/test/depIssuesFixer.spec.ts]: okay, as we discussed we can multiply the variable by some factor

**Pull request merged**